File: core/src/nl/t64/game/rpg/screens/world/GameMap.java
package nl.t64.game.rpg.screens.world;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import com.badlogic.gdx.maps.MapObject;
import com.badlogic.gdx.maps.objects.RectangleMapObject;
import com.badlogic.gdx.maps.tiled.TiledMap;
import com.badlogic.gdx.math.Vector2;
import nl.t64.game.rpg.Utils;
import nl.t64.game.rpg.audio.AudioEvent;
import nl.t64.game.rpg.constants.Constant;
import nl.t64.game.rpg.screens.world.entity.Direction;
import nl.t64.game.rpg.screens.world.pathfinding.TiledGraph;

import java.util.Collection;
import java.util.List;
import java.util.stream.Stream;


public class GameMap {

    public static final int LIGHTMAP_REGION_MULTIPLIER = 10;

    private static final String SOUND_LAYER = "sound";
    private static final String EVENT_LAYER = "event";
    private static final String QUEST_LAYER = "quest";
    private static final String UPPER_TEXTURE_LAYER = "upper_texture";
    private static final String LOWER_TEXTURE_LAYER = "lower_texture";
    private static final String SAVE_LAYER = "save";
    private static final String NPC_LAYER = "npc";
    private static final String HERO_LAYER = "hero";
    private static final String COLLISION_LAYER = "collision";
    private static final String SPAWN_LAYER = "spawn";
    private static final String PORTAL_LAYER = "portal";
    private static final String WARP_LAYER = "warp";
    private static final String REST_LAYER = "rest";
    private static final String LIGHTS_LAYER = "lights";

    private static final String WIDTH_PROPERTY = "width";
    private static final String HEIGHT_PROPERTY = "height";
    private static final String STEP_SOUND_PROPERTY = "step_sound";

    private static final String DEFAULT_STEP_SOUND = "grass";

    final String mapTitle;
    final TiledMap tiledMap;
    final AudioEvent bgm;
    final List<AudioEvent> bgs;
    final List<Texture> lightmapCamera;
    final Sprite lightmapMap;
    final Sprite lightmapPlayer;

    Vector2 playerSpawnLocation;
    Direction playerSpawnDirection;

    final List<GameMapNpc> npcs;
    final List<GameMapHero> heroes;
    final List<GameMapLight> lights;
    final List<GameMapQuestBlocker> questBlockers;
    final List<GameMapQuestTexture> upperTextures;
    final List<GameMapQuestTexture> lowerTextures;
    final List<RectangleMapObject> sparkles;
    final List<RectangleMapObject> chests;
    final List<RectangleMapObject> doors;
    private final List<RectangleMapObject> sounds;
    private final List<GameMapBlocker> blockers;
    private final List<GameMapEvent> eventDiscovers;
    private final List<GameMapQuestDiscoverObject> questDiscovers;
    private final List<GameMapQuestCheckObject> questCheckers;
    private final List<GameMapNote> notes;
    private final List<GameMapSavePoint> savePoints;
    private final List<GameMapSpawnPoint> spawnPoints;
    private final List<GameMapPortal> portals;
    private final List<GameMapWarpPoint> warpPoints;

    private final String defaultStepSound;

    TiledGraph tiledGraph;

    GameMap(String mapTitle) {
        this.mapTitle = mapTitle;
        this.tiledMap = MapManager.getTiledMap(this.mapTitle);
        this.bgm = MapManager.getBgmOfMap(this.tiledMap);
        this.bgs = MapManager.getBgsOfMap(this.tiledMap);

        var loader = new GameMapLayerLoader(tiledMap);

        this.lightmapCamera = loader.loadLightmapCamera();
        this.lightmapMap = loader.loadLightmapMap();
        this.lightmapPlayer = loader.loadLightmapPlayer();
        this.defaultStepSound = this.tiledMap.getProperties().get(STEP_SOUND_PROPERTY, DEFAULT_STEP_SOUND, String.class);

        this.playerSpawnLocation = new Vector2();


        this.sounds = loader.loadLayer(SOUND_LAYER);
        this.npcs = loader.loadLayer(NPC_LAYER, GameMapNpc::new);
        this.heroes = loader.loadLayer(HERO_LAYER, rectObject -> Utils.getGameData().getHeroes().contains(rectObject.getName()), GameMapHero::new);
        this.blockers = loader.loadLayer(COLLISION_LAYER, GameMapBlocker::new);
        this.lights = loader.loadLayer(LIGHTS_LAYER, GameMapLight::new);
        this.eventDiscovers = loader.equalsIgnoreCase(EVENT_LAYER, "discover", GameMapEvent::new);

        this.questBlockers = loader.equalsIgnoreCase(QUEST_LAYER, "blocker", GameMapQuestBlocker::new);
        this.questDiscovers = loader.equalsIgnoreCase(QUEST_LAYER, "discover", GameMapQuestDiscoverObject::new);
        this.questCheckers = loader.equalsIgnoreCase(QUEST_LAYER, "check", GameMapQuestCheckObject::new);

        this.upperTextures = loader.loadTextureLayer(UPPER_TEXTURE_LAYER, GameMapQuestTexture::new);
        this.lowerTextures = loader.loadTextureLayer(LOWER_TEXTURE_LAYER, GameMapQuestTexture::new);

        this.sparkles = loader.startsWith(REST_LAYER, "sparkle");
        this.chests = loader.startsWith(REST_LAYER, "chest");
        this.doors = loader.startsWith(REST_LAYER, "door");
        this.notes = loader.startsWith(REST_LAYER, "note", GameMapNote::new);

        this.savePoints = loader.loadLayer(SAVE_LAYER, GameMapSavePoint::new);
        this.spawnPoints = loader.loadLayer(SPAWN_LAYER, GameMapSpawnPoint::new);
        this.portals = loader.loadLayer(PORTAL_LAYER, rectObject -> new GameMapPortal(rectObject, mapTitle));
        this.warpPoints = loader.loadLayer(WARP_LAYER, rectObject -> new GameMapWarpPoint(rectObject, mapTitle));
    }

    void setTiledGraph() {
        this.tiledGraph = new TiledGraph(getWidth(), getHeight());
    }

    String getUnderground(Vector2 point) {
        return sounds.stream()
                     .filter(underground -> underground.getRectangle().contains(point))
                     .findAny()
                     .map(MapObject::getName)
                     .orElse(defaultStepSound);
    }

    void setPlayerSpawnLocationForNewLoad(String mapTitle) {
        var dummyObject = new RectangleMapObject();
        dummyObject.setName(mapTitle);
        GameMapPortal spawnForNewLoadPortal = new GameMapPortal(dummyObject, mapTitle);
        setPlayerSpawnLocation(spawnForNewLoadPortal);
    }

    void setPlayerSpawnLocation(GameMapRelocator portal) {
        for (GameMapSpawnPoint spawnPoint : spawnPoints) {
            if (spawnPoint.isInConnectionWith(portal)) {
                playerSpawnLocation.set(spawnPoint.getX(), spawnPoint.getY());
                setPlayerSpawnDirection(portal, spawnPoint);
                return;
            }
        }
    }

    private void setPlayerSpawnDirection(GameMapRelocator portal, GameMapSpawnPoint spawnPoint) {
        if (spawnPoint.direction.equals(Direction.NONE)) {
            playerSpawnDirection = portal.enterDirection;
        } else {
            playerSpawnDirection = spawnPoint.direction;
        }
    }

    boolean isOutsideMap(Vector2 point) {
        return point.x < 0 || point.x >= getWidth()
               || point.y < 0 || point.y >= getHeight();
    }

    float getPixelWidth() {
        return getWidth() * (Constant.TILE_SIZE / 2f);
    }

    float getPixelHeight() {
        return getHeight() * (Constant.TILE_SIZE / 2f);
    }

    private int getWidth() {
        return (int) tiledMap.getProperties().get(WIDTH_PROPERTY);
    }

    private int getHeight() {
        return (int) tiledMap.getProperties().get(HEIGHT_PROPERTY);
    }

    void dispose() {
        tiledMap.dispose();
    }

    void debug(ShapeRenderer shapeRenderer) {

        shapeRenderer.setColor(Color.YELLOW);

        Stream.of(blockers, questBlockers)
              .flatMap(Collection::stream)
              .map(GameMapObject::getRectangle)
              .forEach(rect -> shapeRenderer.rect(rect.x, rect.y, rect.width, rect.height));

        shapeRenderer.setColor(Color.BLUE);

        Stream.of(portals, spawnPoints, npcs, heroes, eventDiscovers, questDiscovers, questCheckers, notes)
              .flatMap(Collection::stream)
              .map(GameMapObject::getRectangle)
              .forEach(rect -> shapeRenderer.rect(rect.x, rect.y, rect.width, rect.height));

        Stream.of(sparkles, chests, doors)
              .flatMap(Collection::stream)
              .map(RectangleMapObject::getRectangle)
              .forEach(rect -> shapeRenderer.rect(rect.x, rect.y, rect.width, rect.height));
    }

}

File: core/src/nl/t64/game/rpg/screens/world/GameMap.kt
package nl.t64.game.rpg.screens.world

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.Sprite
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.badlogic.gdx.graphics.glutils.ShapeRenderer
import com.badlogic.gdx.maps.objects.RectangleMapObject
import com.badlogic.gdx.maps.tiled.TiledMap
import com.badlogic.gdx.math.Vector2
import ktx.tiled.*
import nl.t64.game.rpg.Utils.gameData
import nl.t64.game.rpg.audio.AudioEvent
import nl.t64.game.rpg.screens.world.entity.Direction
import nl.t64.game.rpg.screens.world.mapobjects.*
import nl.t64.game.rpg.screens.world.pathfinding.TiledGraph


const val LIGHTMAP_REGION_MULTIPLIER = 10

private const val SOUND_LAYER = "sound"
private const val EVENT_LAYER = "event"
private const val CUTSCENE_LAYER = "cutscene"
private const val QUEST_LAYER = "quest"
private const val UPPER_TEXTURE_LAYER = "upper_texture"
private const val LOWER_TEXTURE_LAYER = "lower_texture"
private const val SAVE_LAYER = "save"
private const val NPC_LAYER = "npc"
private const val HERO_LAYER = "hero"
private const val ENEMY_LAYER = "enemy"
private const val COLLISION_LAYER = "collision"
private const val SPAWN_LAYER = "spawn"
private const val PORTAL_LAYER = "portal"
private const val WARP_LAYER = "warp"
private const val REST_LAYER = "rest"
private const val LIGHTS_LAYER = "lights"

private const val STEP_SOUND_PROPERTY = "step_sound"
private const val DEFAULT_STEP_SOUND = "grass"

class GameMap(val mapTitle: String) {

    val tiledMap: TiledMap = MapManager.getTiledMap(mapTitle)
    val bgm: AudioEvent = MapManager.getBgmOfMap(tiledMap)
    val bgs: List<AudioEvent> = MapManager.getBgsOfMap(tiledMap)

    private val loader = GameMapLayerLoader(tiledMap)

    val parallaxBackground: TextureRegion = loader.loadParallaxBackground()
    val lightmapCamera: List<Texture> = loader.loadLightmapCamera()
    val lightmapMap: Sprite = loader.loadLightmapMap()
    val lightmapPlayer: Sprite? = loader.loadLightmapPlayer()
    private val defaultStepSound: String = tiledMap.property(STEP_SOUND_PROPERTY, DEFAULT_STEP_SOUND)

    lateinit var playerSpawnLocation: Vector2
    lateinit var playerSpawnDirection: Direction
    lateinit var tiledGraph: TiledGraph

    val npcs: List<GameMapNpc> = loader.loadLayer(NPC_LAYER) { GameMapNpc(it) }
    val heroes: List<GameMapHero> = loader.loadLayer(HERO_LAYER, { gameData.heroes.contains(it.name) }, { GameMapHero(it) })
    val enemies: List<GameMapEnemy> = loader.loadLayer(ENEMY_LAYER) { GameMapEnemy(it) }
    val lights: List<GameMapLight> = loader.loadLayer(LIGHTS_LAYER) { GameMapLight(it) }
    val questBlockers: List<GameMapQuestBlocker> = loader.equalsIgnoreCase(QUEST_LAYER, "blocker") { GameMapQuestBlocker(it) }
    val upperTextures: List<GameMapQuestTexture> = loader.loadTextureLayer(UPPER_TEXTURE_LAYER) { GameMapQuestTexture(it) }
    val lowerTextures: List<GameMapQuestTexture> = loader.loadTextureLayer(LOWER_TEXTURE_LAYER) { GameMapQuestTexture(it) }
    val sparkles: List<GameMapSparkle> = loader.startsWith(REST_LAYER, "sparkle") { GameMapSparkle(it) }
    val chests: List<RectangleMapObject> = loader.startsWith(REST_LAYER, "chest")
    val doors: List<RectangleMapObject> = loader.startsWith(REST_LAYER, "door")

    private val sounds: List<RectangleMapObject> = loader.loadLayer(SOUND_LAYER)
    private val blockers: List<GameMapBlocker> = loader.loadLayer(COLLISION_LAYER) { GameMapBlocker(it) }
    private val eventDiscovers: List<GameMapEventDiscover> = loader.equalsIgnoreCase(EVENT_LAYER, "discover") { GameMapEventDiscover(it) }
    private val eventCheckers: List<GameMapEventChecker> = loader.equalsIgnoreCase(EVENT_LAYER, "check") { GameMapEventChecker(it) }
    private val cutsceneDiscovers: List<GameMapCutscene> = loader.equalsIgnoreCase(CUTSCENE_LAYER, "discover") { GameMapCutscene(it) }
    private val questDiscovers: List<GameMapQuestDiscover> = loader.equalsIgnoreCase(QUEST_LAYER, "discover") { GameMapQuestDiscover(it) }
    private val questCheckers: List<GameMapQuestChecker> = loader.equalsIgnoreCase(QUEST_LAYER, "check") { GameMapQuestChecker(it) }
    private val notes: List<GameMapNote> = loader.startsWith(REST_LAYER, "note") { GameMapNote(it) }
    private val savePoints: List<GameMapSavePoint> = loader.loadLayer(SAVE_LAYER) { GameMapSavePoint(it) }
    private val spawnPoints: List<GameMapSpawnPoint> = loader.loadLayer(SPAWN_LAYER) { GameMapSpawnPoint(it) }
    private val portals: List<GameMapPortal> = loader.loadLayer(PORTAL_LAYER) { GameMapPortal(it, mapTitle) }
    private val warpPoints: List<GameMapWarpPoint> = loader.loadLayer(WARP_LAYER) { GameMapWarpPoint(it, mapTitle) }

    val pixelWidth: Float get() = tiledMap.totalWidth().toFloat()
    val pixelHeight: Float get() = tiledMap.totalHeight().toFloat()
    val width: Int get() = tiledMap.width
    val height: Int get() = tiledMap.height

    fun setTiledGraph() {
        tiledGraph = TiledGraph(tiledMap.width, tiledMap.height)
    }

    fun getUnderground(point: Vector2): String {
        return sounds
            .filter { it.rectangle.contains(point) }
            .map { it.name }
            .firstOrNull() ?: defaultStepSound
    }

    fun setPlayerSpawnLocationForNewLoad(mapTitle: String) {
        val dummyObject = RectangleMapObject()
        dummyObject.name = mapTitle
        val spawnForNewLoadPortal = GameMapPortal(dummyObject, mapTitle)
        setPlayerSpawnLocation(spawnForNewLoadPortal)
    }

    fun setPlayerSpawnLocation(portal: GameMapRelocator) {
        spawnPoints.filter { it.isInConnectionWith(portal) }
            .forEach {
                playerSpawnLocation = Vector2(it.x, it.y)
                setPlayerSpawnDirection(portal, it)
                return
            }
    }

    private fun setPlayerSpawnDirection(portal: GameMapRelocator, spawnPoint: GameMapSpawnPoint) {
        playerSpawnDirection = if (spawnPoint.direction == Direction.NONE) {
            portal.enterDirection
        } else {
            spawnPoint.direction
        }
    }

    fun isOutsideMap(point: Vector2): Boolean {
        return point.x < 0 || point.x >= tiledMap.width
                || point.y < 0 || point.y >= tiledMap.height
    }

    fun dispose() {
        tiledMap.dispose()
    }

    fun debug(shapeRenderer: ShapeRenderer) {

        shapeRenderer.color = Color.YELLOW

        listOf(blockers, questBlockers)
            .flatten()
            .map { it.rectangle }
            .forEach { shapeRenderer.rect(it.x, it.y, it.width, it.height) }

        shapeRenderer.color = Color.BLUE

        listOf(
            portals, spawnPoints, npcs, heroes, enemies, eventDiscovers, eventCheckers, cutsceneDiscovers,
            questDiscovers, questCheckers, notes, sparkles
        )
            .flatten()
            .map { it.rectangle }
            .forEach { shapeRenderer.rect(it.x, it.y, it.width, it.height) }

        listOf(chests, doors)
            .flatten()
            .map { it.rectangle }
            .forEach { shapeRenderer.rect(it.x, it.y, it.width, it.height) }
    }

}
